File: LED/LED_source.f90
PROGRAM LED_FirstDWave
IMPLICIT None
! Note:  Function is of the form du/dt+a*du/dx=f(x)
!~  this takes a reconstructed with shockwave form giving a 2nd order accurate values
!Declarations
REAL, ALLOCATABLE, DIMENSION(:,:):: u !wave equation function u
REAL:: a,dphalf,dmhalf,ull,urr,url,ulr,uphalf,umhalf,dx,dt,l,time,x
REAL:: xmh,xph,trapS
INTEGER:: i,j,dxi,dti
OPEN(unit=43,file='LEDOutput.txt')
OPEN(unit=40,file='LEDInput.csv')
!User Input
READ(40,*) ! don't read the first input line
READ(40,*)l,dx,time,dt,a
dxi=int(l/dx+1) ! spatial nodes
dti=int(time/dt+1) ! time nodes
WRITE(*,'(A,I3,A,I3)') 'spatial nodes =',dxi, '   time nodes =',dti
ALLOCATE(u(-1:(dxi+2),dti)) ! add 4 to spatial nodes to get boundary values
READ(40,*) ! don't read this charater line either
!~ READ(40,*) u(1:dxi,1) ! initial values in Input.csv file
READ(40,*) ! Read initial values in Input.csv file as empty line
DO i=1,dxi,1 ! if have function for initial values
	x=real(i-1)*dx ! to calculate the x value on the bar
	u(i,1)=IV(x)
END DO
u(-1:0,:)=0. ! outside boundary values or ghost nodes, left side
u((dxi+1):(dxi+2),:)=0. ! outside boundary values or ghost nodes, right side
WRITE(43,*) u(1:dxi,1)
!Calc
DO j=1,dti-1,1 ! time steps
	DO i=1,dxi,1 ! each dx node on bar (off bar values were noted above)
!LED reconstructed method
! sign(1.,a) to handle negative a values
			urr=u(i+1,j)-sign(1.,a)*dx/2.*(u(i+2,j)-u(i,j))/(2.*dx)
			url=u(i,j)+sign(1.,a)*dx/2.*(u(i+1,j)-u(i-1,j))/(2.*dx)
			ull=u(i-1,j)+sign(1.,a)*dx/2.*(u(i-2,j)-u(i,j))/(2.*dx)
			ulr=u(i,j)-sign(1.,a)*dx/2.*(u(i+1,j)-u(i-1,j))/(2.*dx)
			!no absolute value in dphalf or dmhalf to handle negative a
			dphalf=.5*(a)*(urr-url)
			dmhalf=.5*(a)*(ulr-ull)
			umhalf=.5*abs(a)*(u(i,j)+u(i-1,j))-dmhalf
			uphalf=.5*abs(a)*(u(i,j)+u(i+1,j))-dphalf
! single trapezoidal of source term from x(i-1/2) to x(i+1/2)
			xmh=(real(i-1)-.5)*dx
			xph=(real(i-1)+.5)*dx
			trapS=.5*(xph-xmh)*(source(xph)+source(xmh))
			u(i,j+1)=-a*dt*(uphalf-umhalf)/dx+u(i,j)+trapS*dt/dx
	END DO
	WRITE(43,*) u(1:dxi,j+1)!output
END DO
CLOSE(43)
CLOSE(40)
DEALLOCATE(u)

CONTAINS
FUNCTION source(x) ! this is for the source function and term
IMPLICIT NONE
REAL:: source,x
source=-1./5.*(x-5.)
END FUNCTION
FUNCTION IV(x) ! this is for the initial values if it is a function
IMPLICIT NONE
REAL::IV,x
IV=cos(2.*x)
END FUNCTION


END PROGRAM 

File: LED/C/LED_source.c
// LED_FirstDWave simple
//~ Note:  Function is of the form du/dt+a*du/dx=f(x)
//~  this takes a reconstructed with shockwave form giving a 2nd order accurate values
#include <iostream>//normal other stuff
#include <fstream> //input from file and output to file
#include <math.h>  //sin and cos
#include <string>  //string variables
//~ #include <stdlib.h>//string to double converter atof function
using namespace std;

// user input source and initial value functions
double source(double x) {//source function
	//~ return 0;};//no source term
	return (-1./5.*(x-5.));}; //input function for source term
double IV(double x) {//initial value function
	return(cos(2.*x));};
int main() {
	ofstream output,test;
	ifstream input;
	// input data and read from file
	output.open ("LEDOutput.csv");
	input.open ("LEDInput.csv");
	string A; //dummy string variable
	double l,dx,time,dt,a;
	input>>A>>A>>A>>A>>A;
	cout <<A<<endl;
	input >> l>>dx>>time>>dt>>a;
	cout << "length = "<<l << "  dx = "<<dx << "  time = "<<time
		<< "  dt = "<<dt <<"  a = "<< a << endl;
	// calc using LED method
	int dxi=int(l/dx+1),dti=int(time/dt+1);//spatial and time nodes
	cout<<"spatial nodes = "<<dxi<<"  time nodes = "<<dti<<endl;
	double u[dti][dxi+3];//initialize u array with ghost nodes, 2 on each side of bar
	double urr,url,ull,ulr,dphalf,dmhalf,umhalf,uphalf;//other LED variables
	double xmh,xph,trapS;
	double sofa=(a>-1)?1:-1;//sign of a
	cout<<sofa<<endl;
	for (int i=0;i<dti;i++){//time
		for (int j=0;j<2;j++){//2 left spatial ghost nodes
			u[i][j]=double (0.0);}
		for (int j=dxi+2;j<(dxi+4);j++){//2 right spatial ghost nodes
			u[i][j]=double(0.0);}
			}
	for (int j=2;j<dxi+2;j++){//bar spatial nodes initial values
		u[0][j]=double(IV(double(j-2)*dx));}
	for (int i=0;i<(dti-1);i++) {//time jump
		for (int j=2;j<(dxi+2);j++) {//spatial jump
			urr=u[i][j+1]-sofa*dx/2.*(u[i][j+2]-u[i][j])/(2.*dx);
			url=u[i][j]+sofa*dx/2.*(u[i][j+1]-u[i][j-1])/(2.*dx);
			ull=u[i][j-1]+sofa*dx/2.*(u[i][j-2]-u[i][j])/(2.*dx);
			ulr=u[i][j]-sofa*dx/2.*(u[i][j+1]-u[i][j-1])/(2.*dx);
	//~ no absolute value in dphalf or dmhalf to handle negative a
			dphalf=.5*(a)*(urr-url);
			dmhalf=.5*(a)*(ulr-ull);
			umhalf=.5*fabs(a)*(u[i][j]+u[i][j-1])-dmhalf;
			uphalf=.5*fabs(a)*(u[i][j]+u[i][j+1])-dphalf;
	//~ single trapezoidal of source term from x-1/2 to x+1/2
			xmh=(double (j-2)-.5)*dx;
			xph=(double (j-2)+.5)*dx;
			trapS=.5*(xph-xmh)*(source(xph)+source(xmh));
			u[i+1][j]=-a*dt*(uphalf-umhalf)/dx+u[i][j]+trapS*dt/dx;
		}
	}
	output.precision(17); //output 17 digits of precision
	for (int i=0;i<(dti);i++){
		for (int j=2;j<(dxi+2);j++){
			output<<double(u[i][j])<<" ";
		}
		output<<endl;
	}
	input.close();
	output.close();
return 0;
}
